Compiler/preproc/preprocessor.cpp

```diff
@@ -0,0 +1,429 @@
+#include "preproc/preprocessor.h"
+#include "util/textstreamwriter.h"
+#include "util/textreader.h"
+#include "script/cc_error.h"
+
+#define STRINGIFY2(X) #X
+#define STRINGIFY(X) STRINGIFY2(X)
+
+using namespace AGS::Common;
+
+extern int currentline; // in script/script_common
+
+namespace AGS {
+namespace Preprocessor {
+
+    static const size_t NOT_FOUND = -1;
+
+#if AGS_PLATFORM_OS_WINDOWS
+    static const char * li_end = "\r\n";
+#else
+    static const char * li_end = "\n";
+#endif
+
+    class StringBuilder : TextWriter {
+    private:
+        String _str = nullptr;
+    public:
+        explicit StringBuilder(size_t size) {
+            _str.Reserve(size);
+        }
+
+        void WriteLine(const String& str) override { _str.Append(str); _str.Append(li_end); }
+        bool IsValid() const override { return true; }
+        void WriteChar(char c) override {  _str.AppendChar(c); }
+        void WriteString(const String &str) override { _str.Append(str);}
+        void WriteFormat(const char *fmt, ...) override { va_list argptr; _str.Append(String::FromFormatV(fmt, argptr)); }
+        void WriteLineBreak() override { _str.Append(li_end); }
+
+        String GetString () {return _str;}
+    };
+
+    class StringReader : TextReader{
+    private:
+        String _str = nullptr;
+        size_t _idx = 0;
+    public:
+        StringReader(const String& str) {
+            _str = str;
+            _idx = 0;
+        }
+
+        bool IsValid() const override { return _idx < _str.GetLength(); }
+        char ReadChar() override { return _str.GetAt(_idx++); }
+        String ReadString(size_t length) override { String res = _str.Mid(_idx, length); _idx+=res.GetLength(); return res;}
+        String ReadLine() override {
+            if(_idx >= _str.GetLength()) return nullptr;
+
+            size_t newPos = _str.FindChar('\n', _idx);
+            if(newPos != NOT_FOUND) {
+                size_t length = newPos - _idx;
+                size_t idx = _idx;
+                _idx = newPos+1;
+                if (length == 0) return String("");
+                return _str.Mid(idx,length);
+            } else {
+                return ReadAll();
+            }
+        }
+        String ReadAll() override { size_t idx = _idx; _idx = _str.GetLength(); return _str.Mid(idx, _str.GetLength() - idx); }
+    };
+
+    static bool Contains(std::vector<String> v, const String &s) {
```

Comment:
Formally speaking, here the vector should be passed by a const reference `&` to avoid copying.

Also probably the loop may be replaced with `std::find(v.rbegin(), v.rend(), s) != v.rend();`

Reply:
thanks, this made things really short!

https://github.com/adventuregamestudio/ags/pull/1370/commits/a922cee7f09b669c4b0073e75061cc14048ca579#diff-32411dcb2b7765703f7e4bc2ce6ca1f198d06ede7d9c7e2bcb02612b79c29615R73-R74